File: templates/nodejs-service/service/src/transport/rmqRpcClient.ts
import amqplib from "amqplib";
import { RPCClient } from "./interfaces";
import { rmqConsume, rmqDeclareQueue } from "./rmqHelper";
import { rmqRpcGenerateReplyQueueName, rmqRpcCall } from "./rmqRpcHelper";
import { EnvelopedMessage } from "./envelopedMessage";
import { RmqEventMap } from "./rmqEventMap";


export class RmqRPCClient implements RPCClient {

    constructor(private logger: Console, private connection: amqplib.Connection, private eventMap: RmqEventMap) {}

    public async call(functionName: string, ...inputs: any[]): Promise<any> {
        const self = this;
        const exchangeName = self.eventMap.getExchangeName(functionName);
        const queueName = self.eventMap.getQueueName(functionName);
        return new Promise(async (resolve, reject) => {
            try {
                const replyTo = rmqRpcGenerateReplyQueueName(queueName);
                const ch = await self.connection.createChannel();
                // consume
                await rmqDeclareQueue(ch, replyTo, {durable: true, autoDelete: true});
                rmqConsume(ch, replyTo, {durable: true, autoDelete: true}, true, async (rmqMessageOrNull) => {
                    try {
                        const rmqMessage = rmqMessageOrNull as amqplib.ConsumeMessage;
                        const jsonEnvelopedOutput = rmqMessage.content.toString();
                        const envelopedOutput = new EnvelopedMessage(jsonEnvelopedOutput);
                        if (envelopedOutput.getErrorMessage()) {
                            self.logger.log(`[ERROR RmqRPCClient] Get Error Reply ${functionName}`, JSON.stringify(inputs), ":", envelopedOutput.getErrorMessage());
                            return reject(new Error(envelopedOutput.getErrorMessage()));
                        }
                        const message = envelopedOutput.getMessage();
                        self.logger.log(`[INFO RmqRPCClient] Get Reply ${functionName}`, JSON.stringify(inputs), ":", JSON.stringify(message.output));
                        resolve(message.output);
                    } catch (err) {
                        self.logger.error(`[ERROR RmqRPCClient] Error While Processing Reply ${functionName}`, JSON.stringify(inputs), ":", err);
                        reject(err);
                    }
                });
                // timeout
                const timeout = self.eventMap.getRpcTimeout(functionName);
                setTimeout(() => {
                    const err = new Error(`Timeout ${timeout}`);
                    self.logger.error(`[ERROR RmqRPCClient] Get timeout ${functionName} ${inputs}: ${err}`)
                    reject(err);
                }, timeout);
                // send message
                self.logger.log(`[INFO RmqRPCClient] Call ${functionName}`, JSON.stringify(inputs));
                await rmqRpcCall(ch, exchangeName, replyTo, inputs);
            } catch (err) {
                self.logger.error(err);
                reject(err);
            }
        });
    }

}

File: templates/python-service/service/transport/rmqRpcClient.py
from typing import Any, Dict, List, Tuple, cast
from pika.adapters.blocking_connection import BlockingConnection, BlockingChannel
from pika.spec import Basic, BasicProperties
from logging import Logger
from .interfaces import Message, RPCClient
from .rmqRpcHelper import rmq_rpc_generate_reply_queue_name, rmq_rpc_call
from .rmqHelper import rmq_consume, rmq_declare_queue, OnMessageCallback
from .rmqEventMap import RmqEventMap
from .envelopedMessage import EnvelopedMessage

import time


class RmqRPCClient(RPCClient):

    def __init__(self, logger: Logger, connection: BlockingConnection, event_map: RmqEventMap):
        self.connection: BlockingConnection = connection
        self.logger: Logger = logger
        self.event_map: RmqEventMap = event_map

    def call(self, function_name: str, *inputs: Any) -> Any:
        # reply state
        reply_state: Dict[str, Any] = self._create_default_reply_state()
        exchange_name = self.event_map.get_exchange_name(function_name)
        queue_name = self.event_map.get_queue_name(function_name)
        # send message
        reply_to = rmq_rpc_generate_reply_queue_name(queue_name)
        ch = self.connection.channel()
        rmq_declare_queue(ch, reply_to, True, True, {})
        rmq_consume(ch, reply_to, True, True, True, {}, self._create_reply_handler(function_name, inputs, reply_state))
        self.logger.info(
            "[INFO RmqRPCClient] Call {} {}".format(function_name, inputs))
        rmq_rpc_call(ch, exchange_name, reply_to, cast(List[Any], inputs))
        # waiting
        start = time.time() * 1000
        timeout = self.event_map.get_rpc_timeout(function_name)
        while not reply_state["accepted"]:
            self.connection.process_data_events()
            if start + timeout < time.time() * 1000:
                reply_state["error_message"] = "Timeout {}".format(timeout)
                self.logger.info("[ERROR RmqRPCClient] Get timeout {} {}: {} ms".format(function_name, inputs, timeout))
                break
        # return or throw error
        if reply_state["error_message"] != "":
            raise Exception(reply_state["error_message"])
        return reply_state["output"]

    def _create_default_reply_state(self) -> Dict[str, Any]:
        return {"accepted": False, "output": None, "error_message": ""}

    def _create_reply_handler(self, function_name: str, inputs: Tuple, reply_state: Dict[str, Any]) -> OnMessageCallback:
        def on_reply(ch: BlockingChannel, method: Basic.Deliver, properties: BasicProperties, json_enveloped_output: str) -> Any:
            if reply_state["accepted"]:
                return
            reply_state["accepted"] = True
            try:
                enveloped_output = EnvelopedMessage(json_enveloped_output)
                if enveloped_output.error_message:
                    self.logger.info("[ERROR RmqRPCClient] Get Error Reply {} {}: {}".format(
                        function_name, inputs, enveloped_output.error_message))
                    reply_state["error_message"] = enveloped_output.error_message
                    return
                self.logger.info("[INFO RmqRPCClient] Get Reply {} {}: {}".format(
                    function_name, inputs, enveloped_output.message))
                message = enveloped_output.message
                reply_state["output"] = message["output"]
            except Exception as e:
                self.logger.info("[ERROR RmqRPCClient] Error While Processing Reply {} {}: {}".format(
                    function_name, inputs, e))
                reply_state["error_message"] = str(e)
        return on_reply 